### ISSUE TITLE
[RFC] Performer Image Categorization

### ISSUE THREAD

**Issue by AdultSun** [stashapp/stash-box#237]
## Scope
There is a significant desire to better organize performer images on stash-box, particularly by allowing images to be categorized by crop, pose, and state of dress.

## Problem
Stash-Box performer images are currently organized by resolution only, basically serving the highest resolution image first followed by the rest in descending order. More popular performers can have many images and our StashDB guidelines encourage a diverse set for each performer representing various crops, poses, states of dress, and general "looks" from the performer's career. Typically, users will prefer a particular type of performer image for their personal use and are often left navigating stash-box's image carousel one-by-one until they find the "best" image. Currently, the only way to set a default image for a particular performer in stash-box is by making sure the preferred image has a higher resolution then the rest of the set. This is an imperfect work-around that requires maintaining a hard cap on image resolution for each performer rather than a true solution.

##  Proposal
The implementation of image categories could help to alleviate these issues. If individual images can be tagged with categories, users could set their own preferred defaults within both stash-box and their personal stash. A convenient system for categorization is already in use in the [actress-images repo](https://gitgud.io/Trizkat/actress-pics/-/tree/master) hosted by Trizkat. Though the collection is exclusively female performers (outside of a small number of AFAB non-binary performers), their system appears to be neutral enough to be suitable for performers of all genders. [In their own words](https://gitgud.io/Trizkat/actress-pics/-/blob/master/SCHEMA.md), with some light editing: 

### **Crop**
How much of the performer is visible?
- **Face**: Above the bust, and up
- **Bust**: Somewhere between the navel and bust, and up
- **Torso**: Above hips, below navel
- **3Q**: Three-quarters view, about mid-thigh, and up
- **Sit**: When you see the person sitting
- **Kneel**: Where you can see the knees touch the ground
- **Wide**: For landscape images where the subject is in some horizontal position
### **Pose**
How is the performer positioned in relation to the camera?
- **Portrait**
  - **Front**: Mostly facing towards the camera
  - **Back**: Mostly facing away from the camera
  - **Side**: Mostly facing perpendicular to the camera
- **Landscape**
  - **Back**: On their back
  - **Fours**, On all fours
  - **Front**: Mainly a frontal view and not covered by previous
### **State of Dress**
What is the performer wearing, if anything?
- **Non-Nude**: Covered up, or can't be determined
- **Topless**: Defined by if areola and nipple is uncovered
- **Nude**: Exposed bust and pubic region
### **Set**
Which photo set is this image from?
- The actress-images schema assigns each image a "set" number in their filename, grouping images of various crops, poses, and states that were originally part of the same photoshoot. This allows users to easily find a matching set of images for use in applications other than Stash that allow for multiple performer images, such as P-Vault.

Possible expansions on this system could include a full-length crop for head-to-toe images, more specifics for non-nude states (swimsuit/lingerie/casual/formal), and approximate release dates/years to allow users to sort by date. Many more recent uploads to the actress-images repo represent a slightly expanded 3Q crop with more specific guidelines for the model's positioning within the frame, which may or may not necessitate an additional category of 3Q+ or similar.

## Standardization
Similar to other areas of stash-box, easily visible guidelines on relevant edit pages will greatly aid less experienced editors in conforming to established community standards. In this case, those guidelines would be clear definitions for the newly introduced terms above. This will require an agreed-upon level of standardization for these images to encourage consistency from performer to performer. Questions to consider while developing this standard:
- **How specific should our standard be?** The borrowed definitions above are fairly general, referencing vague areas like "about mid-thigh" for the bottom of 3/4 portraits. If a greater level of consistency is preferred, we would need to define the positioning of performers within the frame with more specificity. Providing image templates that include literal guidelines for each crop may be necessary.
- **How big should the images be?** This includes aspect ratio, resolution, and file size. The current standards for StashDB are laid out in our [shared guidelines doc](https://docs.google.com/document/d/18IlkSIhHf4QA9ulxPHExARZGs0FmNg8O-4TYTdZL4Iw/edit#heading=h.3wxd2adg0xf0) as a 2:3 ratio for vertical/portrait images, resolutions up to 666 x 1000, and file sizes of around 120-160 KB. The doc also notes that the "maximum size for images is 1280 pixels" and that anything larger will be "automatically reduced to that size." Images that are considered "low quality" in terms of content, resolution, or size are recommended to be deleted assuming higher quality images are available, though only a few specific examples of "low quality" are established. Including more "official" guidelines within StashDB itself would provide an opportunity to revise these standards if necessary. Images should be high enough in resolution and quality to meet most users preferences, but with file sizes small enough to not cause issues with hosting space in stash-box, inflated database sizes in stash, or slow loading speeds in a browser.
- **How closely should our standard align with the actress-images repo?** Their current standards are listed as 2:3 and "at least 1200px high" for vertical/portrait, 16:9 and 1280x720px for horizontal/landscape, around 350 kB maximum file size "though not at the cost of lower quality", and a recommended standard JPEG export described as "roughly 80% quality (very high)". However, more recent contributions to their repo are significantly higher than this standard, with some images at a resolution of 1275x1875 with sizes approaching 2 MB. Their minimum standards are already slightly higher than our maximum, so some compromise between the two standards may allow for a greater level of collaboration between the two projects.

## Further Development
In addition to a more basic implementation of categories for performer images, more advanced features have been suggested:
- An image editor within the browser could allow users to conform to more specific standards without the need of advanced standalone software like Photoshop or GIMP. This embedded image editor could allow for basic operations like cropping the image with visible guidelines of the community-standard, downscaling or compressing to an acceptable size, and possibly even upscaling.
- If our standards are aligned closely enough with those of the actress-images repo, an automatic import process could be possible as well. Their file naming schema is specific enough to allow for parsing of the performer name, crop, pose, state of dress, and image set number. An automated system could theoretically detect additions to the repo, adjust the size of the new image if necessary, and submit it to the proper performer on stash-box for approval after assigning the relevant categories. If all images in stash-box are given a perceptual hash, it could detect duplicates that are already saved in stash-box as well. Then it could either submit the new image as a replacement or cancel the submission altogether as appropriate.

**Comment by echo6ix** [comment]
Well done @AdultSun 

First, I think we can borrow a lot of pointers for our guidelines from https://www.themoviedb.org/bible/image

>  easily visible guidelines on relevant edit pages will greatly aid less experienced editors in conforming to established community standards

Yes. Very important. I am lazy and often neglect reading rules/guidelines unless they're in-my-face, clear, and concise. I am probably not alone here. Rub the guidelines in my face on the same page where I am uploading an image.

> How big should the images be?

I am not opposed to images larger than 1280 pixels, but I also have not read a compelling reason why they need to be larger than that. These images will be populating the performer's profile picture in Stash, they aren't being consumed in the image galleries, so what is the point?

In any case, if it is decided to use higher resolution images, consider implementing an automated process that creates an alternate resized version (1280 pixels) of every high-res image. Users could have a local Stash setting that toggles between downloading hi-res or low-res performer images. This would satisfy the concerns Fonzie@Discord mentioned re: high-res images.

> An image editor within the browser could allow users to conform to more specific standards without the need of advanced standalone software like Photoshop or GIMP...

If y'all suspect a large majority of StashDB image content is going to be manually crowdsourced, then do this. I do not think an image editor per se is a required, but a Javascript image cropper with hard-coded guide overlays that correspond to each image "category".

I strongly suspect this will leave less open to user interpretation and result in higher quality and more consistent image crop contributions.

> How specific should our standard be?

High level of specificity for consistent content.

This is a sample from my own personal guidelines that I stole from a modeling photography site:

## Portrait head shot types

### 1/4 tight headshot
Crop from the collar bone area or just above shoulders[^shoulder] to just above the head.[^head] [Link to headshot template finalized by AdultSun and I](https://discord.com/channels/559159668438728723/798641040029777980/1036122884126548088).

~~### 1/4 headshot
Crop from mid-section, generally just below the the breast to just above the head.[^head]~~

### 1/2 headshot (bust)
Crop from the bellybutton (or precisely just above it as it's a great fixed reference point) to just above the head.[^head]

### 3/4 headshot
Crop from just above the knee to just above the head.[^head]

## Body shot types

Incomplete because I don't care about full body shots :D

[^shoulder]: Crop just above the shoulder bone, that is clip out the shoulder bones, but keep the trapezius.
[^head]: Always leave a small amount of padding between the top of the head and the top of image.

**Comment by AdultSun** [comment]
Thank you for your feedback, echo6ix! That page from TheMovieDB looks like a good resource. The professional definitions you found there are helpful too, and already line up nicely with the actress-images terms. One quick clarification...

> I do not think an image editor per se is a required, but a Javascript image cropper with hard-coded guide overlays that correspond to each image "category".

That section was me half-remembering your suggestion from Discord a while back. You can blame my lack of technical expertise if my explanation was misleading. I try to use my words carefully but sometimes the correct terminology escapes me, so please feel free to question/correct me when that happens.

I agree that cropping images with guide overlays is really the only important action a feature like that would need to do. The only other functionality that could be helpful there is basic color adjustments or resizing/compressing to fit resolution/file size guidelines, but I imagine the latter would be implemented in the backend somewhere instead.

**Comment by spincity07** [comment]
@AdultSun nice work on this RFC. I like the flexibility it provides and can see a day where you set the cover image you prefer in Stash and it pulls it from Stashdb for that or all performers. 

I'd also agree that file sizes don't need to be that big. This will save space and bandwidth as well, but the performer images aren't very big in stash. No need to keep them gigantic on the server. 

> I agree that cropping images with guide overlays is really the only important action a feature like that would need to do.

I couldn't agree with this more. Hit everyone over the head with the guidelines and make it idiot proof. 

I have nothing more to add but I wanted to affirm as an outside voice that I like this direction.

**Comment by Chodbin** [comment]
I would love to see this implemented as well with a couple of further suggestions;

**Local Stash**
If the images are categorized, it would be nice if you could choose your preference when using tagger, for example;

**State of Dress**

- Preference 1 - Topless
- Preference 2 - Nude
- Preference 3 - Non-Nude

**Pose**

- Preference 1 - Portrait
- Preference 1a - Front
- Preference 1b - Side
- Preference 1c - Back
- Preference 2 - Landscape
- Preference 2a - Front
- Preference 2b - Back
- Preference 2c - Fours

**Crop**

- Preference 1 - Bust
- Preference 2 - Torso
- Preference 3 - 3Q
- Preference 4 - Face
- Preference 5 - Sit
- Preference 6 - Kneel
- Preference 7 - Wide

So if these were my preferences it would always try and find first
Topless, Portrait, Front, Bust

**Scenes**

Something similar could be implemented for scene images, but simpler.
At the moment a scene can only have 1 image, I was thinking people could submit up to 3 categorized images in the following priority for stashdb (which a user could override in their local stash with preferences on a per-studio basis)

- Official Cover/Poster - Landscape
- Official Cover/Poster - Portrait
- Screenshot - Landscape

Perhaps also when people submit an image, a warning that it is a lower resolution than existing if they are replacing an image. There are probably some instances where this might be wanted, but some kind of confirmation/reason for it would be good, e.g. Existing image only shows performers foot

**Comment by stg-annon** [comment]
related to cropping these libraries might be useful:

[smartcrop.js](https://github.com/jwagner/smartcrop.js) could be useful for automatically enforcing the standards that we want to use


[JavaScript-Load-Image](https://github.com/blueimp/JavaScript-Load-Image) would be more of an in browser editor for manual use

**Comment by echo6ix** [comment]
My updated opinion. If people want high-quality (and consistent looking) performer assets I think this aspect of stash-box would need to almost become a project unto itself or a sub-section of stash-box that doesn't democratically function like the rest of it.

I say this based on my experience with [fanart.tv](https://fanart.tv/) and StashDB, and the fact that Stash users are beginning to create their own performer art projects, such as performer clearart (high quality performer images with background transparency). Eventually they'll want to submit them to a repository.

- Moderators with expertise to approve or reject submissions (a democratic voting based system will not ensure high-quality because reviewers need and intimate understanding of all the rules. See [fanart.tv](https://fanart.tv/).
- Moderators/admins with that same expertise to administrate the project
- Very specific unambiguous rules for submissions and categories such as (how much padding is required by the edges of images, cropping rules, text/watermark removal rules, etc)
- Tutorials, guides, and templates
- Dedicated section on Discourse including a request channel (where users request images of specific performers and styles)
- Limitation of submissions per user per performer per category

Basically, look at what [fanart.tv](https://fanart.tv/) has done and incorporate elements of that.

**Comment by discourse-stashapp** [comment]
This issue has been mentioned on **Stash**. There might be relevant details there:

https://discourse.stashapp.cc/t/cropping-guide-templates/4560/1

**Comment by Maista6969** [comment]
I want to try implementing this feature and I'd like to get some feedback on some of the thoughts I've had reading this RFC before I start writing any code

First I'd like to talk about the prior art I've reviewed, primarily from Musicbrainz since their [Cover Art Archive](https://musicbrainz.org/doc/Cover_Art_Archive) is a valuable source of inspiration for this design with one major difference: the images in CAA are _evidence_ but performer images are _utility_: a CAA scan documents a physical object that may no longer exist and cannot be re-photographed, so it's value is as a faithful record so a flawed scan or watermarked image is worth labeling as such and keeping. 

For performers in stash-box the usage is different, these images are often cropped or replaced by better ones as time goes on and their primary purpose (in my opinion) is to definitely identify a performer by documenting what their face and body looks like.

Note that this is different for scene covers, but that's a different kettle of fish and our design should be able to support their peculiarities as well.

### Categories

My first thought was to make the categories editable (like site types, for example) but I don't think that's nearly as applicable here: the reason site types are editable is that different stash-boxes can have different priorities in what links they'd like to allow and how they'd like them to be sorted, which is sort of like performer images, but as we've seen in this discussion there are already widely applicable definitions for what kinds of images a performer could have. This suggests a closed taxonomy that can be used by every stash-box, which has the major advantage that anyone interacting with multiple stash-boxes can use the same definitions.

If these image types were to be customizable per instance it would have a much bigger cost for the user: if Stash is to build support for image preferences when scraping, the types _must_ be consistent across stash-box endpoints. A client cannot offer "prefer headshots" without first fetching the type vocabulary from each server and asking the user to manually map them out.

Another advantage of a closed taxonomy is that the edit records stay readable: since edits are stored as JSON a label like `{"image_type": 3}` is worthless as soon as an admin edits the definitions, but a value like `{"image_type": "CROP_FACE"}` stays readable forever and is portable to future software.

A third and final advantage of having a closed taxonomy for shot and crop types is that we can ship templates for these, both for users to download and use locally but also for an in-browser cropping tool.

I'd err on the side of having too many types than too few: removing a type would be a Herculean task (imagine removing a site-type like "Studio Profile" on StashDB) but adding a new type is trivial.

With that in mind, I'm planning on images being categorized by types that are _grouped_ in orthogonal dimensions, so we can have things like "Headshot"/"Three quarters"/"Full body" and "Non-nude"/"Lingerie"/"Nude" ranked independently. These groups would be exclusive (an image can be either "headshot" or a "full body", not both) and of course optional for backwards compatibility. A fully classified image would be labeled with one type per group, like "Portrait" because it's a posed photo of the performer, "Bust" because of how it's cropped, "Front" because of the performer facing and "Topless" because of the state of dress.

The proposed type group for shot types, "what type of photograph is this?"

| type     | Meaning                                                                                                                                                       |
| -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Portrait | A photograph of the performer, posed or promotional, often from a picture gallery                                                                             |
| Candid   | Unposed, like event photography, behind-the-scenes, or just screenshots from a video                                                                          |
| Detail   | A close-up of a feature rather than the person: tattoo, piercing, scar, [third testicle](https://stashdb.org/performers/eb25ae9e-4e3c-4e8d-8169-fd564fe6897b) |

This group would be the first determinant when ranking images because a close-up of a tattoo should never become a performer's primary image, no matter how well-cropped it is

The proposed type group for crop, "how much of them is in frame?"

| Type               | Meaning                                                                                                                    |
| ------------------ | -------------------------------------------------------------------------------------------------------------------------- |
| Face               | Collarbone up                                                                                                              |
| Bust               | Navel or chest up                                                                                                          |
| Three quarter      | Mid-thigh up                                                                                                               |
| Three quarter plus | Looser three-quarter, framed on 80-20 thirds |
| Full body          | Head to toe                                                                                                                |
| Torso              | Hips to shoulders, head not necessarily included                                                                           |
| Wide               | Subject horizontal or small in frame                                                                                       |

The proposed type group for pose, "where is the camera?", is just "Front", "Side" and "Back"

The proposed type group for state of dress, "what are they wearing", is just "Non-nude", "Swimwear/Lingerie", "Topless" and "Nude"

3 shot types x 7 crop types x 3 pose types x 4 dress types means there are __252 different ways to completely classify an image__.

One idea I really liked is the suggestion from @echo6ix that some images have objectively better utility in certain parts of the UI, so I'd like to add a "thumbnail" field to the performer GraphQL type that would default to a headshot. This would be used in search dropdowns and the like.

### What I'd leave out

Image voting: "What is the best image?" is a very subjective question and I think we should try to avoid these as much as possible because it's essentially just noise. I'd rather give users the ability to order their own preferences both for scraping and browsing.

Orientation: whether an image is landscape or portrait can be derived from its dimensions, no need to label this manually.

"Other" as a valid choice in any group: in a fixed taxonomy this would be actively harmful, an information-free bucket that hides the fact that an image is essentially uncategorized.

Themed sets, or sets in general: open-ended in a way that the rest of the taxonomy is not, prone to arguments about what should and should not be included. The war on Christmas never ends.

TMDB and fanart.tv have categories like "Poster", "Backdrop", "Profile" etc. because they designate what part of a page the image is suitable for: this is something that Stash does not currently support (outside of plugins) so I don't think it should be a category at this point.

Career eras: since this varies wildly I'd rather add this capability as a date that can optionally be attached to an image, an approximate date like both scenes and performers support in stash-box.

### Future work

The design should be extensible to scene cover images (see #648) and even studio images (see #279)

I'd like to implement cropping tools (with templates) to the frontend to help users submit images that conform better to the categories we have defined, but that's probably better as a separate PR

Once the types have been deployed to stash-boxes and we're happy with how they're working out we could certainly deploy machine-learning tools to help categorize performer images through bot edits

**Comment by AdultSun** [comment]
I agree with all of your conclusions here. I still think image voting has its utility in a couple scenarios — default sorting when no user preference is set, or breaking ties when a user's preference ranks them the same — but that functionality can wait for later. I also agree that the cropping tools and templates can wait as well.

Another dimension of categories that we've used before in Trizkat's old repo is Pose (standing, kneeling, squatting, sitting, on fours, prone, supine) separate from simple Orientation (front, side, back). [Trizkat's schema](https://gitgud.io/Trizkat/actress-pics/-/blob/master/SCHEMA.md) has those bundled into the Crop category, which makes some sense since they're often either implied (3Q = Standing), mutually exclusive (Full body ≠ Sitting), or irrelevant (Face). Those sitting and kneeling portraits are pretty common too, though I don't know how many users would rank them higher than standing. Still, I think it's worth considering if we want a closed but extensive taxonomy for all stash-boxes to share. Categorizing them would allow us to come up with cropping templates for them too.

Another thought I had about the "closed" list is that even if we don't allow individual stash-boxes to add their own custom categories (outside of forking the repo I guess), those stash-box admins should still be able to turn off unwanted categories. I think that would also help us expand the centralized list more freely if each stash-box doesn't feel forced to subscribe to everything.

**Comment by Maista6969** [comment]
> Those sitting and kneeling portraits are pretty common too, though I don't know how many users would rank them higher than standing. Still, I think it's worth considering if we want a closed but extensive taxonomy for all stash-boxes to share. Categorizing them would allow us to come up with cropping templates for them too.

I'd be happy to add more poses, like I mentioned it's easier to have more than we need than less than we need: if an image type ends up being redundant it can trivially be merged with another. I'm already planning on using Trizkat's actress-pics as a test corpus for my plans 🙂

> Another thought I had about the "closed" list is that even if we don't allow individual stash-boxes to add their own custom categories (outside of forking the repo I guess), those stash-box admins should still be able to turn off unwanted categories. I think that would also help us expand the centralized list more freely if each stash-box doesn't feel forced to subscribe to everything.

This is a great point and one I'll be incorporating into the design!

**Comment by echo6ix** [comment]
> we can ship templates for these, both for users to download and use locally but also for an in-browser cropping tool.

Thanks for doing this. Please borrow the well-established ux pattern from passport id croppers, where a rudimentary SVG template outline is overlayed on the cropping selection.

Example
- https://www.idphoto4you.com/?Target=CropPage

I've probably submitted thousands of tight headshots using a template and it just makes the consistency almost idiot proof.

**Comment by InfiniteStash** [comment]
I was all prepared to argue for custom image categories, but I agree with your points.

> Career eras: since this varies wildly I'd rather add this capability as a date that can optionally be attached to an image, an approximate date like both scenes and performers support in stash-box.

I think this is a good idea to get in, so we can start capturing data as soon as possible. Doesn't need to be required or anything, just nice to have.

I'm wondering if we should require certain aspect ratios from the jump or not. Having a standardized 2:3 size would make them a lot more useful in stash. The initial PR doesn't have to include cropping tools, but if we're going to require aspect ratios it should be a requirement from the start.

**Comment by DogmaDragon** [comment]
> I'm wondering if we should require certain aspect ratios from the jump or not. Having a standardized 2:3 size would make them a lot more useful in stash. The initial PR doesn't have to include cropping tools, but if we're going to require aspect ratios it should be a requirement from the start.

I think requiring it without built-in tools would lead to worse outcomes. It just adds too much friction to adding images if people are expected to modify their images in a 3rd party application.

**Comment by InfiniteStash** [comment]
> > I'm wondering if we should require certain aspect ratios from the jump or not. Having a standardized 2:3 size would make them a lot more useful in stash. The initial PR doesn't have to include cropping tools, but if we're going to require aspect ratios it should be a requirement from the start.
> 
> I think requiring it without built-in tools would lead to worse outcomes. It just adds too much friction to adding images if people are expected to modify their images in a 3rd party application.

I agree cropping would have to be part of the same release, I'm just suggesting doing that work in a separate PR.

**Comment by AdultSun** [comment]
My main concern is if we start requiring a strict aspect ratio before including the standardized templates it will lead to a lot of wasted effort. Now every image is cropped once _without_ a template, just to be cropped again later _with_ a template. And if an image was cropped oddly the first time, it might not be easy or possible to crop it again to match the template.

I've always imagined the standardized stuff as a checkbox parallel to the other image categories, but still tied into the same user-defined ranking system. So it would also feel weird to start standardizing images based on aspect ratio alone, without any means for users to prioritize them, and to not consider those images truly "standardized" later on because they don't conform to the templates.

It leaves them in an odd middle-ground that I'm not sure is worth it. If we want to encourage editors to standardize everything, then we should fully standardize them in one go.

**Comment by Maista6969** [comment]
I've already started exploring the templates/cropping design and here are some thoughts:

- Like labels, cropping is __optional__ but encouraged: this sort of feature is already going to attract a certain kind of user and I'm sure the capability will be enough to improve the quality of new images that are added without actually forcing anything. In some situations there genuinely will not be possible to get a good image for every crop and we'll have to accept sub-par images at least until the performer has better source material available. Think of the commonly exploited college girls and how they often start out with hastily cropped screenshots before they get more work done.
- The cropping should not be done in the browser to avoid compatibility issues, but in the backend. stash-box already ships with libvips which can do rotation/cropping cheaply and efficiently, guaranteeing consistent results regardless of what the browser would do.
- The guidelines/templates should live in the backend and they should be .psd so there's only one source of truth. This way the overlay can be derived from the template and we can offer the exact same template for download that the stash-box uses internally. I'd also love to make them live right on disk so we can easily swap them out (if we get better templates) or delete (for instances that do not care for such precision). This also makes it easier to add different templates for uses that Stash does not currently have, like a template for banner images

Here's a mockup based on an image I used in a [recent edit](https://stashdb.org/edits/01a02013-dd41-71f1-ac6f-f33cc1e9e2e9) that I would have definitely cropped better if it was easier to do

<img width="648" height="994" alt="Image" src="https://github.com/user-attachments/assets/bb2c2d69-1eb6-4401-b931-e2430a1dbb60" />

This would appear when uploading images to encourage people to crop, and it would then send the full image + the rotation/crop specification to the server where it would be applied and saved.

I am basing this on the templates in the aforementioned [actress-pics](https://gitgud.io/Trizkat/actress-pics/) repository, but I'd love it if we could find someone with professional experience (maybe in the community?) to make the exact templates we'd like, with named guide lines and everything, so that we could license them cleanly

**Comment by AdultSun** [comment]
> I am basing this on the templates in the aforementioned [actress-pics](https://gitgud.io/Trizkat/actress-pics/) repository, but I'd love it if we could find someone with professional experience (maybe in the community?) to make the exact templates we'd like, with named guide lines and everything, so that we could license them cleanly

I'm not a professional, but I created and uploaded those actress-pics templates myself. So those are at least the exact templates *I* would like. 
https://gitgud.io/Trizkat/actress-pics/-/commit/7198eb5f17603b1c95b0ceb89fbe929373d683cd

I don't know what that means for licensing though since I uploaded them to a different repo. I think I originally discussed and uploaded them to Stash's Discord before that if it matters, but I'd have to look to make sure.

**Comment by AdultSun** [comment]
> This would appear when uploading images to encourage people to crop, and it would then send the full image + the rotation/crop specification to the server where it would be applied and saved.

I like the idea of saving a copy of the original image too, by the way. That's always been useful with Stash-Box's method of saving the full scene image, but serving a smaller copy by default unless you specifically ask for the original. Stash-Box has historically had trouble with draft submissions attempting to replace the full-size original with its own downscaled copy though, so that's something to watch out for here too.

**Comment by Maista6969** [comment]
> > I am basing this on the templates in the aforementioned [actress-pics](https://gitgud.io/Trizkat/actress-pics/) repository, but I'd love it if we could find someone with professional experience (maybe in the community?) to make the exact templates we'd like, with named guide lines and everything, so that we could license them cleanly
> 
> I'm not a professional, but I created and uploaded those actress-pics templates myself.

Oh wow, I guess I should have checked the commit history! I'm sure you'll be fine with me adding them to the stash-box repo then, where they'd end up under the MIT license like the rest of it?

**Comment by AdultSun** [comment]
Yeah, you have my permission. I actually don't see a license in Trizkat's repo either so I guess that means reusing them is fine?

**Comment by echo6ix** [comment]
Some feedback because I've ironed out a lot of this in the past when revising my headshot uploads.

I used to use eyes as a hard anchor and abandoned it as a guide for all images except those when the subject is square to the camera. I learned that using facial anchors will create very inconsistent head sizes unless the subject is square to the camera, and they rarely are.

The above image is a perfect example. The template calls for a collarbone-up crop. It anchors a bisection of the eyes, but the angle of the subject invalidates the collarbone rule and forces the crop through the neck. If the subject was closer to square to the camera both rules could be met, but it would make the size of the head smaller.

We're dealing with images that will have high variance with the subject's angles, creating high variance in cropped head sizes. To mitigate that you prioritize the size of the head within a bounding box. That's why I suggested using an oval or face outline in the cropper. It will lean towards more consistent head sizes for even lazy end-users using the cropper.

The second goal to consider is UI extensibility. Ensuring the size of the head is consistent creates a safe zone on the bottom 1/4 of the image that can be used for any future UI overlays.

```
Top 3/4 (The Face)
- Center the head from any angle within this boundary
- Use top of the head to the bottom of the chin as internal anchors
- A visual head or oval outline in the cropper tool will guide users to hit this mark with
  less open for interpretation

Bottom 1/4 (The Safe Zone)
- Captures the neck down
- Guarantees a persistent UI safe zone for any future designs that call for UI overlays (performer name, etc)
```

Using the 3/4 : 1/4 rule I can add overlays without obstructing the subject's face:

<img width="215" height="309" alt="Image" src="https://github.com/user-attachments/assets/77d7adfe-b446-4240-8de4-7d2c69c90a04" />

Just something to consider before standardizing a template.
